From d84030a5275c6e9d82cefda4dc032134d2b66b5b Mon Sep 17 00:00:00 2001
From: Robert S Peterson <rpeterso@redhat.com>
Date: Thu, 11 Sep 2014 14:22:37 -0400
Subject: [fs] gfs2: Hash the negative dentry during inode lookup

Message-id: <374796646.20832327.1410445357298.JavaMail.zimbra@redhat.com>
Patchwork-id: 93771
O-Subject: [RHEL7.1 PATCH][bz#1123986] GFS2: Hash the negative dentry during inode lookup
Bugzilla: 1123986
RH-Acked-by: Steven Whitehouse <swhiteho@redhat.com>
RH-Acked-by: Abhijith Das <adas@redhat.com>

Hi,

This patch is a cherry-pick from this upstream patch:
http://git.kernel.org/cgit/linux/kernel/git/steve/gfs2-3.0-nmw.git/commit/fs/gfs2?id=7b7a91152d3c2ddca95172cac1675625cc8dffaf

A kernel was built to include this patch, here:
https://brewweb.devel.redhat.com/taskinfo?taskID=7947691

It was tested on systems gfs-i8c-01.mpc.lab.eng.bos.redhat.com (nfs server)
and gfs-i24c-01.mpc.lab.eng.bos.redhat.com (nfs client) using reliable
failing scenarios documented in the bugzilla.

Patch description:

Fix a regression introduced by:
6d4ade986f9c8df31e68 GFS2: Add atomic_open support
where an early return misses d_splice_alias() which had been
adding the negative dentry.

rhbz#1123986

Signed-off-by: Benjamin Coddington <bcodding@redhat.com>
Signed-off-by: Bob Peterson <rpeterso@redhat.com>
Signed-off-by: Steven Whitehouse <swhiteho@redhat.com>
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/fs/gfs2/inode.c b/fs/gfs2/inode.c
index 7b03d30..d45bad5 100644
--- a/fs/gfs2/inode.c
+++ b/fs/gfs2/inode.c
@@ -772,8 +772,10 @@ static struct dentry *__gfs2_lookup(struct inode *dir, struct dentry *dentry,
  int error;
 
  inode = gfs2_lookupi(dir, &dentry->d_name, 0);
- if (!inode)
+ if (inode == NULL) {
+  d_add(dentry, NULL);
   return NULL;
+ }
  if (IS_ERR(inode))
   return ERR_CAST(inode);
 
-- 
1.7.1